Enable resource update aliasing in auto-clustering

trentlo opened the pull request:
We see cases where enabling resource update aliasing greatly reduces memory footprints.

trentlo committed:
[XLA/GPU] Enable resource update in auto-clustering.

trentlo commented:
@cheshire could you help to take a look when you have a moment? Thanks!

trentlo committed:
[XLA/GPU] Polish comments.

cheshire commented on tensorflow/compiler/jit/kernels/xla_ops.cc:
```diff
@@ -419,11 +419,16 @@ void XlaCompileOp::Compute(OpKernelContext* ctx) {
-    // Do not alias resource updates as locking variables in XlaCompile and
-    // unlocking them in XlaRun may lead to deadlocks.
+    // We turn on the `may_alias_resource_update` using the following strategy.
+    // In XalRunOp, if we observe the resource variables have been written, we
```
XlaRunOp

cheshire commented on tensorflow/compiler/jit/kernels/xla_ops.cc:
```diff
@@ -503,13 +508,26 @@ void XlaRunOp::Compute(OpKernelContext* ctx) {
+      ctx, *closure.compilation_result(), closure.num_constant_args());
+  OP_REQUIRES_OK(ctx, variable_infos.status());
+  OP_REQUIRES_OK(ctx, LockVariables(absl::MakeSpan(*variable_infos)));
+  std::map<int, const Tensor*> resource_var_ptrs;
```
absl::flat_hash_map normally preferred

trentlo replied:
Will do.

cheshire commented on tensorflow/compiler/jit/kernels/xla_ops.cc:
```diff
@@ -520,11 +538,25 @@ void XlaRunOp::Compute(OpKernelContext* ctx) {
+      const Tensor* tensor_ptr = nullptr;
+      if (p.second.has_value()) {
+        if (resource_var_ptrs.count(p.first) &&
+            resource_var_ptrs[p.first]->SharesBufferWith(p.second.value())) {
```
Could you expand why the logic here is different from the one in XlaLocalLaunchOp? Maybe we could unify these two?

trentlo replied:
The difference is due to an additional issue in auto-clustering that the compilation and run are in two ops (`XlaRunOp` and `XlaCompileOp`). So, a snapshot of tensors is stored by `XlaCompileOp` to maintain the shape consistency. The snapshotted tensors need to be used as inputs if the shapes of the `Vars` change. In contrast, snapshot is not needed for manual clustering.

Here, we check if some snapshotted tensors actually are the copies of the tensors in `Vars`. If yes, we release the snapshotted tensors (i.e., copies of the Var tensors) and in effect the Var tensors' ref counts will decrement by one. This will expose the opportunities for `tensor_donation`. In contrast, if we don't release the snapshotted tensors, `tensor_donation` will never happen because the ref counts will at least be 2 (i.e., the copy due to snapshot _and_ the tensor stored in the Var).

Note that the snapshotted tensors will not be copies of the Var tensors if the Vars are written.

Overall, `XlaCompileOp`/`XlaRunOp` need to deal with the snapshotted tensors for shape consistency. If the shape actually do not change and the underlying tensors are the same, we want the buffer donation to happen.

trentlo reviewed (commented):
Responded. Will push commits soon. Please help to take a look again.

trentlo committed:
[XLA/GPU] Code change to address review comments.

trentlo commented:
Just check if there is anything that needs to be taken care of for this PR? It is fine if it simply waits for merging. Thanks!

cheshire commented:
For some reason it's failing TPU tests internally. Let me take a look.

cheshire commented:
Is it possible at all to think about tests for this? We have tests measuring GPU memory consumption in `def_function_xla_jit_test`. WDYT about a test which runs in autoclustering configuration, runs a trivially clusterable resource var update, and then asserts that we have not consumed additional RAM?

cheshire commented on tensorflow/compiler/jit/xla_launch_util.cc:
```diff
@@ -254,7 +254,8 @@ XlaComputationLaunchContext::PopulateInputs(
         is_resource_variable &&
         absl::c_any_of(compilation_result->resource_updates,
                        [&](const XlaCompiler::ResourceUpdate& update) {
-                         return update.input_index == i && update.modified;
```
Could you clarify why `i` was changed to `arg_num`?

trentlo replied:
Because `resource_updates` stores `arg_num` (see [here](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/compiler/tf2xla/xla_compiler.cc;l=302?q=file:tensorflow%2Fcompiler%2Ftf2xla%2Fxla_compiler.cc)), which is an input index seen at the compilation time.

`i` refers to the input shapes of the compiled kernel. `arg_num` and `i` can differ in the auto-clustering mode. They might be no different for manual clustering.

cheshire replied:
> They might be no different for manual clustering

Can we CHECK it then?

trentlo reviewed (commented):
Response to the `arg_num` is inlined.

Adding a test to assert some GPU memory invariance sounds a good idea to me. Is the `get_total_memory_usage('GPU')` accurate enough and not interfered by `BFCAllocator`? 

BTW, is the internal TPU failure related to this PR?

cheshire commented:
> BTW, is the internal TPU failure related to this PR?

Unfortunately yes. We get different numerics. No idea why.

>  Is the get_total_memory_usage('GPU') accurate enough and not interfered by BFCAllocator?

Cf. def_function_xla_jit_test. It is accurate, BFCAllocator is disabled for tests IIRC.

trentlo commented:
Thanks @gbaned for launching the CI test again!

@cheshire, could you help me to take a quick look at whether the previous numerical different persists in this new CI results? If yes, I can try separating the `arg_num` change into another PR with necessary CHECKs.

gbaned commented:
@trentlo Can you please check @cheshire's comments and keep us posted ? Thanks!

trentlo commented:
> @trentlo Can you please check @cheshire's comments and keep us posted ? Thanks!

I'll see if I can separate the arg_num change into another PR and give it some tests. Creating test for it may not be easy.

trentlo commented:
So, I have separated the arg_num change into another PR and it's merged.

Now, I'm trying to add a python test for this PR, but I don't know what is the best way to trigger XLA AUTO_JIT in the python test?

I added a test as:
```
 tf_xla_py_test(
    name = "def_function_auto_jit_test",
    srcs = ["def_function_auto_jit_test.py"],
    use_xla_device = True,
    deps = [
        ":def_function",
        "//tensorflow/compiler/tests:xla_test",
        "//tensorflow/python:constant_op",
    ],
)
```

However, the AUTO_JIT is not triggered:

`2021-09-21 22:41:33.498184: I tensorflow/compiler/jit/mark_for_compilation_pass.cc:1678] Not compiling cluster with device /job:localhost/replica:0/task:0/device:GPU:0`

@cheshire, could you give me some guidance about how to trigger XLA AUTO_JIT for  `tf_xla_py_test` (or even `cuda_py_test`)?

cheshire commented:
I believe def_function_xla_jit_test has *_xla_gpu autoclustering variant, e.g. tested here: https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/python/eager/def_function_xla_jit_test.py;l=44-68?q=file:def_function_xla_jit_test

trentlo commented:
> I believe def_function_xla_jit_test has *_xla_gpu autoclustering variant, e.g. tested here: https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/python/eager/def_function_xla_jit_test.py;l=44-68?q=file:def_function_xla_jit_test

Right, I did check that one for reference.

`test_util.is_xla_enabled()` is always `false` in that test though. I did not figure out how to enable it.

cheshire commented:
It's false if you call (e.g.) def_function_xla_jit_test_cpu, can you run
def_function_xla_jit_test_xla_gpu? Speaking of, nearly every TF test should
have an _xla_gpu variant which runs autoclustering.

On Tue, Sep 21, 2021 at 5:34 PM Trent Lo ***@***.***> wrote:

> I believe def_function_xla_jit_test has *_xla_gpu autoclustering variant,
> e.g. tested here:
> https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/python/eager/def_function_xla_jit_test.py;l=44-68?q=file:def_function_xla_jit_test
>
> Right, I did check that one for reference.
>
> test_util.is_xla_enabled() is always false in that test though. I did not
> figure out how to enable it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/51353#issuecomment-924488092>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACVGHYEKI4YJV4YBJ4LBS3UDEP7XANCNFSM5BWPKQ4Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

trentlo commented:
Like this?

```
root@sc-sdgx-420:/opt/tensorflow/tensorflow-source# CUDA_VISIBLE_DEVICES=3 bazel test --jobs=20 --test_output=all --cache_test_results=no -c opt --config=cuda --cxxopt=-D_GLIBCXX_USE_CXX11_ABI=0 --output_filter MATCH_NOTHING --javabase=@bazel_tools//tools/jdk:remote_jdk11 //tensorflow/python/eager:def_function_xla_jit_test_xla_gpu
...
ERROR: no such target '//tensorflow/python/eager:def_function_xla_jit_test_xla_gpu': target 'def_function_xla_jit_test_xla_gpu' not declared in package 'tensorflow/python/eager' (did you mean 'def_function_xla_jit_test_gpu'?) defined by /opt/tensorflow/tensorflow-source/tensorflow/python/eager/BUILD
INFO: Elapsed time: 1.968s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (2 packages loaded)
FAILED: Build did NOT complete successfully (2 packages loaded)
```

trentlo commented:
> Like this?
> 
> ```
> root@sc-sdgx-420:/opt/tensorflow/tensorflow-source# CUDA_VISIBLE_DEVICES=3 bazel test --jobs=20 --test_output=all --cache_test_results=no -c opt --config=cuda --cxxopt=-D_GLIBCXX_USE_CXX11_ABI=0 --output_filter MATCH_NOTHING --javabase=@bazel_tools//tools/jdk:remote_jdk11 //tensorflow/python/eager:def_function_xla_jit_test_xla_gpu
> ...
> ERROR: no such target '//tensorflow/python/eager:def_function_xla_jit_test_xla_gpu': target 'def_function_xla_jit_test_xla_gpu' not declared in package 'tensorflow/python/eager' (did you mean 'def_function_xla_jit_test_gpu'?) defined by /opt/tensorflow/tensorflow-source/tensorflow/python/eager/BUILD
> INFO: Elapsed time: 1.968s
> INFO: 0 processes.
> FAILED: Build did NOT complete successfully (2 packages loaded)
> FAILED: Build did NOT complete successfully (2 packages loaded)
> ```

For some reasons, I can only succeed with `def_function_xla_jit_test_gpu`, i.e., with the `_gpu` postfix. Do I miss some configurations?

trentlo committed:
[XLA/GPU] add a unittest for resource var update in Autoclustering.

trentlo commented:
I add a test case here:
https://github.com/tensorflow/tensorflow/pull/51353/files#diff-4eb0ebadc5b54abae1338833f7cf762f309c1386e897c3674361a76cb20e8266R28

It appears that I have to add `context.context().optimizer_jit = True` to enable the Autoclustering.

trentlo committed:
[XLA/GPU] Fine tune def_function_auto_jit_test.

Fine tune the parameters such that the test is deemed to fail if resource var
update does not occur.

trentlo commented:
Add one more test to test the signature including compile-time constant, similar to the one tested in `tf.function(jit_compile=True)`.

Please take a look again. Thanks!

trentlo committed:
[XLA/GPU] add testUpdateVariableWithCompileTimeConstMemoryUsage()

testing a signature of (compile-time const, args, res_var). The compile-time
const will be optimized away so that the kernel signature will become
(args, res_var).

trentlo commented:
"Windows Bazel GPU" failed with the following message. Will try fixing it.

```
==================== Test output for //py_test_dir/tensorflow/python/eager:def_function_auto_jit_test_gpu:
T:\tmp\bigvaudl\execroot\org_tensorflow\bazel-out\x64_windows-opt\bin\tensorflow\tools\ci_build\gpu_build\parallel_gpu_execute: line 35: is_absolute: command not found
Running test T:/tmp/bigvaudl/execroot/org_tensorflow/bazel-out/x64_windows-opt/bin/py_test_dir/tensorflow/python/eager/def_function_auto_jit_test_gpu.exe --test_device=XLA_GPU --types=DT_HALF,DT_FLOAT,DT_DOUBLE,DT_UINT8,DT_QUINT8,DT_INT8,DT_QINT8,DT_INT32,DT_QINT32,DT_INT64,DT_BOOL,DT_COMPLEX64,DT_COMPLEX128,DT_BFLOAT16 on GPU 1
Traceback (most recent call last):
  File "\\?\T:\tmp\Bazel.runfiles_m80h9ijg\runfiles\org_tensorflow\py_test_dir\tensorflow\python\eager\def_function_auto_jit_test.py", line 20, in <module>
    from tensorflow.compiler.tests import xla_test
ModuleNotFoundError: No module named 'tensorflow.compiler.tests'
```

cheshire commented:
If it's just Windows, and if there's a way to disable the test on Windows, it would be fine.

trentlo commented:
> If it's just Windows, and if there's a way to disable the test on Windows, it would be fine.

Thanks for the tip. I copied and pasted the `tags` from `def_function_xla_jit_test`.

trentlo committed:
[XLA/GPU] Add tags to pass CI for def_function_auto_jit_test_gpu.

trentlo commented:
> > If it's just Windows, and if there's a way to disable the test on Windows, it would be fine.
> 
> Thanks for the tip. I copied and pasted the `tags` from `def_function_xla_jit_test`.

@cheshire, gentle ping~

gbaned commented:
Here are the internal errors, @trentlo can you please verify ?

Traceback (most recent call last):
  File "/tensorflow/compiler/tests/adadelta_test.py", line 114, in testBasic
    rtol=1e-5)
  File "/tensorflow/python/framework/test_util.py", line 1446, in decorated
    return f(*args, **kwds)
  File "/tensorflow/python/framework/test_util.py", line 3068, in assertAllCloseAccordingToType
    self.assertAllClose(a, b, rtol=rtol, atol=atol, msg=msg)
  File "/tensorflow/python/framework/test_util.py", line 1446, in decorated
    return f(*args, **kwds)
  File "/tensorflow/python/framework/test_util.py", line 3019, in assertAllClose
    self._assertAllCloseRecursive(a, b, rtol=rtol, atol=atol, msg=msg)
  File "/tensorflow/python/framework/test_util.py", line 2981, in _assertAllCloseRecursive
    (path_str, path_str, msg)))
  File "/tensorflow/python/framework/test_util.py", line 2913, in _assertArrayLikeAllClose
    a, b, rtol=rtol, atol=atol, err_msg="\n".join(msgs), equal_nan=True)
  File "/py/numpy/testing/_private/utils.py", line 1528, in assert_allclose
    verbose=verbose, header=header, equal_nan=equal_nan)
  File "/py/numpy/testing/_private/utils.py", line 840, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Not equal to tolerance rtol=0.001, atol=0.001
Mismatched value: a is different from b. 
not close where = (array([0, 1]),)
not close lhs = [0.005707 0.005707]
not close rhs = [0.003883 0.003883]
not close dif = [0.001823 0.001823]
not close tol = [0.001004 0.001004]
dtype = float16, shape = (2,)
Mismatched elements: 2 / 2 (100%)
Max absolute difference: 0.001823
Max relative difference: 0.4695
 x: array([0.005707, 0.005707], dtype=float16)
 y: array([0.003883, 0.003883], dtype=float16)

trentlo commented:
> 0.001823

Thanks for the info.

This looks like a reasonable loss due to finite numerical precision to me. abs=0.001 is too restricted for fp16. Let me see if I can repro the behavior with the OSS build.

gbaned commented:
@trentlo Any update on this PR? Please. Thanks!

trentlo commented:
> @trentlo Any update on this PR? Please. Thanks!

Thanks for checking.

I tested (>10 times) with `python adadelta_test.py --test_device XLA_GPU --types=DT_HALF` but I don't repro the issue somehow. May I know whether the test failed on GPU or TPU in your CI?

gbaned commented:
@cheshire Can you please assist on above comments from @trentlo. Thanks!

cheshire commented:
> but I don't repro the issue somehow. May I know whether the test failed on GPU or TPU in your CI?

Maybe just bump up the tolerance and retry?

trentlo commented:
> > but I don't repro the issue somehow. May I know whether the test failed on GPU or TPU in your CI?
> 
> Maybe just bump up the tolerance and retry?

I can't exclude the possibility that there is indeed a bug somewhere after I inspected the logs in your CI and the relevant test codes. So, it is risky to simply dump up the tolerance.

I tend to believe that this particular test (adadelta_test.py) may be written in a way that causes strange interactions between XLA and TF with to the PR. It is also possible that the test itself is not well written but I can't be sure. A data point is that the codes in this PR are already in our internal TF fork for quite a while, and it seems fine.

Let me close this PR for now and reopen later if needed. Thanks for all the replies and comments.
